mypy/dmypy_server.py

```diff
@@ -557,6 +572,14 @@ def fine_grained_increment(
             # Use the remove/update lists to update fswatcher.
             # This avoids calling stat() for unchanged files.
             changed, removed = self.update_changed(sources, remove or [], update or [])
+        if explicit_export_types:
+            # If --export-types is given, we need to force full re-checking of all
+            # explicitly passed files, since we need to visit each expression.
+            changed += [
+                (bs.module, bs.path)
+                for bs in sources
+                if bs.path and (bs.module, bs.path) not in changed
```

Comment:
This is quadratic, and it could be slow if there are many files and many that are changed. Maybe construct a temporary set from `changed` and use it in the `not in changed` check?

Reply:
Yes, good point, fixed now.